File: src/main/java/com/github/jactor/persistence/entity/BlogEntity.java
package com.github.jactor.persistence.entity;

import static java.util.Objects.hash;
import static java.util.stream.Collectors.toSet;

import com.github.jactor.persistence.dto.BlogDto;
import com.github.jactor.persistence.dto.PersistentDto;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Stream;
import javax.persistence.AttributeOverride;
import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.Embedded;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.Table;
import javax.validation.constraints.NotNull;
import org.apache.commons.lang3.builder.ToStringBuilder;
import org.apache.commons.lang3.builder.ToStringStyle;

@Entity
@Table(name = "T_BLOG")
public class BlogEntity implements PersistentEntity<BlogEntity> {

  @Id
  private Long id;

  @Embedded
  @AttributeOverride(name = "createdBy", column = @Column(name = "CREATED_BY"))
  @AttributeOverride(name = "timeOfCreation", column = @Column(name = "CREATION_TIME"))
  @AttributeOverride(name = "modifiedBy", column = @Column(name = "UPDATED_BY"))
  @AttributeOverride(name = "timeOfModification", column = @Column(name = "UPDATED_TIME"))
  private PersistentDataEmbeddable persistentDataEmbeddable;

  @Column(name = "CREATED")
  private LocalDate created;
  @Column(name = "TITLE")
  private String title;
  @JoinColumn(name = "USER_ID")
  @ManyToOne(cascade = CascadeType.MERGE, fetch = FetchType.LAZY)
  private UserEntity userEntity;
  @OneToMany(mappedBy = "blog", fetch = FetchType.EAGER, cascade = CascadeType.MERGE)
  private Set<BlogEntryEntity> entries = new HashSet<>();

  @SuppressWarnings("unused")
  BlogEntity() {
    // used by entity manager
  }

  private BlogEntity(BlogEntity blogEntity) {
    created = blogEntity.created;
    entries = blogEntity.entries.stream().map(BlogEntryEntity::copyWithoutId).collect(toSet());
    id = blogEntity.id;
    persistentDataEmbeddable = new PersistentDataEmbeddable();
    title = blogEntity.getTitle();
    userEntity = Optional.ofNullable(blogEntity.getUser()).map(UserEntity::copyWithoutId).orElse(null);
  }

  public BlogEntity(@NotNull BlogDto blogDto) {
    created = blogDto.getCreated();
    id = blogDto.getId();
    persistentDataEmbeddable = new PersistentDataEmbeddable();
    title = blogDto.getTitle();
    userEntity = Optional.ofNullable(blogDto.getUser()).map(UserEntity::new).orElse(null);
  }

  public BlogDto asDto() {
    return new BlogDto(
        initPersistentDto(),
        created, title, Optional.ofNullable(userEntity).map(UserEntity::asDto).orElse(null)
    );
  }

  public void add(BlogEntryEntity blogEntryEntity) {
    blogEntryEntity.setBlog(this);
    entries.add(blogEntryEntity);
  }

  @Override
  public BlogEntity copyWithoutId() {
    BlogEntity blogEntity = new BlogEntity(this);
    blogEntity.setId(null);

    return blogEntity;
  }

  @Override
  public PersistentDto initPersistentDto() {
    return new PersistentDto(getId(), getCreatedBy(), getTimeOfCreation(), getModifiedBy(), getTimeOfModification());
  }

  @Override
  public Stream<PersistentEntity> streamSequencedDependencies() {
    return Stream.concat(streamSequencedDependencies(userEntity), entries.stream());
  }

  @Override
  public boolean equals(Object o) {
    return this == o || o != null && getClass() == o.getClass() &&
        Objects.equals(getTitle(), ((BlogEntity) o).getTitle()) &&
        Objects.equals(getUser(), ((BlogEntity) o).getUser());
  }

  @Override
  public int hashCode() {
    return hash(getTitle(), getUser());
  }

  @Override
  public String toString() {
    return new ToStringBuilder(this, ToStringStyle.SHORT_PREFIX_STYLE)
        .appendSuper(super.toString())
        .append(getCreated())
        .append(getTitle())
        .append(getUser())
        .toString();
  }

  @Override
  public String getCreatedBy() {
    return persistentDataEmbeddable.getCreatedBy();
  }

  @Override
  public LocalDateTime getTimeOfCreation() {
    return persistentDataEmbeddable.getTimeOfCreation();
  }

  @Override
  public String getModifiedBy() {
    return persistentDataEmbeddable.getModifiedBy();
  }

  @Override
  public LocalDateTime getTimeOfModification() {
    return persistentDataEmbeddable.getTimeOfModification();
  }

  @Override
  public Long getId() {
    return id;
  }

  @Override
  public void setId(Long id) {
    this.id = id;
  }

  public LocalDate getCreated() {
    return created;
  }

  public Set<BlogEntryEntity> getEntries() {
    return entries;
  }

  public String getTitle() {
    return title;
  }

  public UserEntity getUser() {
    return userEntity;
  }

  public void setTitle(String title) {
    this.title = title;
  }

  public static BlogEntity aBlog(BlogDto blogDto) {
    return new BlogEntity(blogDto);
  }
}

File: persistence/src/main/kotlin/com/github/jactor/persistence/entity/BlogEntity.kt
package com.github.jactor.persistence.entity

import java.time.LocalDate
import java.time.LocalDateTime
import java.util.Objects
import org.apache.commons.lang3.builder.ToStringBuilder
import org.apache.commons.lang3.builder.ToStringStyle
import com.github.jactor.persistence.dto.BlogDto
import jakarta.persistence.AttributeOverride
import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Embedded
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.OneToMany
import jakarta.persistence.SequenceGenerator
import jakarta.persistence.Table

@Entity
@Table(name = "T_BLOG")
class BlogEntity : PersistentEntity<BlogEntity?> {
    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "blogSeq")
    @SequenceGenerator(name = "blogSeq", sequenceName = "T_BLOG_SEQ", allocationSize = 1)
    override var id: Long? = null

    @Embedded
    @AttributeOverride(name = "createdBy", column = Column(name = "CREATED_BY"))
    @AttributeOverride(name = "timeOfCreation", column = Column(name = "CREATION_TIME"))
    @AttributeOverride(name = "modifiedBy", column = Column(name = "UPDATED_BY"))
    @AttributeOverride(name = "timeOfModification", column = Column(name = "UPDATED_TIME"))
    private lateinit var persistentDataEmbeddable: PersistentDataEmbeddable

    @Column(name = "CREATED")
    var created: LocalDate? = null
        private set

    @Column(name = "TITLE")
    var title: String? = null

    @JoinColumn(name = "USER_ID")
    @ManyToOne(cascade = [CascadeType.PERSIST, CascadeType.MERGE], fetch = FetchType.LAZY)
    var user: UserEntity? = null

    @OneToMany(mappedBy = "blog", fetch = FetchType.EAGER, cascade = [CascadeType.PERSIST, CascadeType.MERGE])
    private var entries: MutableSet<BlogEntryEntity> = HashSet()

    constructor() {
        // used by entity manager
    }

    private constructor(blogEntity: BlogEntity) {
        created = blogEntity.created
        entries = blogEntity.entries
            .map { obj: BlogEntryEntity -> obj.copyWithoutId() }
            .toMutableSet()
        id = blogEntity.id
        persistentDataEmbeddable = PersistentDataEmbeddable()
        title = blogEntity.title
        user = blogEntity.user?.copyWithoutId()
    }

    constructor(blogDto: BlogDto) {
        created = blogDto.created
        id = blogDto.id
        persistentDataEmbeddable = PersistentDataEmbeddable(blogDto.persistentDto)
        title = blogDto.title
        user = blogDto.userInternal?.let { UserEntity(it) }
    }

    fun asDto(): BlogDto {
        return BlogDto(
            persistentDataEmbeddable.asPersistentDto(id), created, title, user?.asDto()
        )
    }

    fun add(blogEntryEntity: BlogEntryEntity) {
        blogEntryEntity.blog = this
        entries.add(blogEntryEntity)
    }

    override fun copyWithoutId(): BlogEntity {
        val blogEntity = BlogEntity(this)
        blogEntity.id = null
        return blogEntity
    }

    override fun modifiedBy(modifier: String): BlogEntity {
        persistentDataEmbeddable.modifiedBy(modifier)
        return this
    }

    override fun equals(other: Any?): Boolean {
        return this === other || other != null && javaClass == other.javaClass &&
            title == (other as BlogEntity).title &&
            user == other.user
    }

    override fun hashCode(): Int {
        return Objects.hash(title, user)
    }

    override fun toString(): String {
        return ToStringBuilder(this, ToStringStyle.SHORT_PREFIX_STYLE)
            .appendSuper(super.toString())
            .append(created)
            .append(title)
            .append(user)
            .toString()
    }

    override val createdBy: String
        get() = persistentDataEmbeddable.createdBy
    override val timeOfCreation: LocalDateTime
        get() = persistentDataEmbeddable.timeOfCreation
    override val modifiedBy: String
        get() = persistentDataEmbeddable.modifiedBy
    override val timeOfModification: LocalDateTime
        get() = persistentDataEmbeddable.timeOfModification

    fun getEntries(): Set<BlogEntryEntity> {
        return entries
    }

    companion object {
        @JvmStatic
        fun aBlog(blogDto: BlogDto): BlogEntity {
            return BlogEntity(blogDto)
        }
    }
}
